### DISCUSSION TITLE
Raw Data Available from Federal CERP Website?

### DISCUSSION THREAD

**Discussion by super25k** [fgpv-vpgf/fgpv-vpgf#4132] [Q&A] [answered]
Hi,

Does anyone know if the raw data from the Federal CERP website (https://atlas.gc.ca/cerp-rpep/en/) is available for download, either by API for otherwise?
Not sure if this is the right place to ask - but I figured someone would know more than me on the topic.

Some of the data can be downloaded by exporting data from the data tables, but not all.
In particular I'd like to get the biomass resource potential data. I can download the ObjectID's and values, but it does not provide a polyline, or any geospatial reference data to go with it.

Thanks in advance.
Rgs,
Mike

**Comment by james-rae** [comment, marked as answer]
Good day Mike.

So this repo is for the software used to power the map on that website, but the website itself and data sources are managed by a different department altogether (Natural Resources Canada).

However I did a bit of snooping on the traffic for that site and can provide an option that may be useful; otherwise you'll need to sniff out a contact for the owner of that site.

The site appears to be getting its data from [this ArcGIS server](https://geoappext.nrcan.gc.ca/arcgis/rest/services/Energy/CERP_EN/MapServer/).

You can make API calls against this server to extract the data it holds.
As an example, [this url](https://geoappext.nrcan.gc.ca/arcgis/rest/services/Energy/CERP_EN/MapServer/25/query) will let you generate a query to extract data from the first sublayer of the "Biomass Resource Potential" (layer index 26 is the second one) . I find setting `1=1` on the "where" field, and `*` in the "Out Fields" field will give me everything. At the bottom is a Format selector, so you may want something like JSON. 

Note these servers usually have a return count maximum, so check the bottom of the result for an `"exceededTransferLimit": true` message. This means there is more data beyond the limit (usually 1000 records), so you might have to start crafting more queries with the where clause being something like `OBJECTID > 1000` and adjusting the number based on the previous result maximum.

So hopefully you're a bit of a programmer and this is easy as pie. You can look at the URL that gets crafted from that query page and start just adjusting it manually (or writing some calls in code) if that is easier.

**Reply by super25k** [reply]
Fantastic, thankyou @james-rae!
I've got a python script to extract ArcGIS REST data so that should work ok.
Thanks for the tips on extraction.
May I ask - what method/tools did you use to sniff out the ArcGIS server? I would like to learn about that, would come in very handy on other data.
Many thanks.

**Reply by james-rae** [reply]
Just used my trusty pal `F12` and went to the Network tab. Cleared it to erase all the loading noise, then turned on the layer. Can see the request asking the server for the layer data.

```
https://geoappext.nrcan.gc.ca/arcgis/rest/services/Energy/CERP_EN/MapServer/export?dynamicLayers=[{"id":25,"name":"Oven-dried cereal straw (tonnes per year)","source":{"type":"mapLayer","mapLayerId":25},"drawingInfo":{"transparency":0},"minScale":0,"maxScale":10000}]&dpi=96&transparent=true&format=png32&layers=show:25&bbox=-7959905.537846077,-1462879.0245110495,5668809.219583441,4077507.056261113&bboxSR=3978&imageSR=3978&size=1717,698&f=image
```

Then just took the url up to `MapServer` to get the root endpoint.
Can be a bit trickier if layers load upfront, you then need to sift through the mountain of startup calls to find sources.
Alternate approaches is to look for `.json` files being downloaded, as these can contain the configuration for the app and have the layer URLs encoded. But a `json` file is not always guaranteed to be the configuration source.

**Reply by super25k** [reply]
Excellent - thanks for the detail!
